File: utilities/java/CredentialDecryptor.java
package com.credentialcode.utility;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import org.json.JSONObject;

/**
 * A utility class for decrypting credentials from .creds files
 */
public class CredentialDecryptor {
    private final byte[] key;
    
    /**
     * Initialize with a base64-encoded key string
     */
    public CredentialDecryptor(String keyBase64) {
        this.key = Base64.getDecoder().decode(keyBase64.trim());
    }
    
    /**
     * Initialize with key bytes
     */
    public CredentialDecryptor(byte[] keyBytes) {
        this.key = keyBytes;
    }
    
    /**
     * Initialize by reading key from file
     */
    public CredentialDecryptor(Path keyPath) throws IOException {
        String keyBase64 = Files.readString(keyPath).trim();
        this.key = Base64.getDecoder().decode(keyBase64);
    }
    
    /**
     * Initialize by reading key from file (convenience method)
     */
    public CredentialDecryptor(File keyFile) throws IOException {
        this(keyFile.toPath());
    }
    
    /**
     * Decrypt a single credential
     */
    public String decrypt(JSONObject encryptedData) throws Exception {
        // Decode from base64
        byte[] ciphertext = Base64.getDecoder().decode(encryptedData.getString("data"));
        byte[] nonce = Base64.getDecoder().decode(encryptedData.getString("nonce"));
        byte[] tag = Base64.getDecoder().decode(encryptedData.getString("tag"));
        
        // Create cipher
        SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec gcmSpec = new GCMParameterSpec(128, nonce);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec);
        
        // Combine ciphertext and tag for Java's GCM implementation
        byte[] combined = new byte[ciphertext.length + tag.length];
        System.arraycopy(ciphertext, 0, combined, 0, ciphertext.length);
        System.arraycopy(tag, 0, combined, ciphertext.length, tag.length);
        
        // Decrypt
        byte[] decrypted = cipher.doFinal(combined);
        
        return new String(decrypted, StandardCharsets.UTF_8);
    }
    
    /**
     * Load and decrypt all credentials from a .creds file
     */
    public Map<String, String> loadCredentials(String credsPath) throws Exception {
        return loadCredentials(Paths.get(credsPath));
    }
    
    /**
     * Load and decrypt all credentials from a .creds file
     */
    public Map<String, String> loadCredentials(Path credsPath) throws Exception {
        String jsonContent = Files.readString(credsPath);
        return loadCredentialsFromJson(jsonContent);
    }
    
    /**
     * Load and decrypt all credentials from a .creds file
     */
    public Map<String, String> loadCredentials(File credsFile) throws Exception {
        return loadCredentials(credsFile.toPath());
    }
    
    /**
     * Load and decrypt all credentials from JSON string
     */
    public Map<String, String> loadCredentialsFromJson(String jsonString) throws Exception {
        JSONObject credsData = new JSONObject(jsonString);
        JSONObject credentials = credsData.getJSONObject("credentials");
        
        Map<String, String> result = new HashMap<>();
        
        for (String key : credentials.keySet()) {
            JSONObject encryptedData = credentials.getJSONObject(key);
            result.put(key, decrypt(encryptedData));
        }
        
        return result;
    }
    
    /**
     * Load and decrypt all credentials from JSONObject
     */
    public Map<String, String> loadCredentialsFromJson(JSONObject credsData) throws Exception {
        JSONObject credentials = credsData.getJSONObject("credentials");
        
        Map<String, String> result = new HashMap<>();
        
        for (String key : credentials.keySet()) {
            JSONObject encryptedData = credentials.getJSONObject(key);
            result.put(key, decrypt(encryptedData));
        }
        
        return result;
    }
    
    /**
     * Convenience method to load credentials with default key path
     */
    public static Map<String, String> loadCredentials(String credsPath) throws Exception {
        Path keyPath = Paths.get(".credential-code", "encryption-key.txt");
        CredentialDecryptor decryptor = new CredentialDecryptor(keyPath);
        return decryptor.loadCredentials(credsPath);
    }
    
    /**
     * Convenience method to load credentials with custom key path
     */
    public static Map<String, String> loadCredentials(String credsPath, String keyPath) throws Exception {
        CredentialDecryptor decryptor = new CredentialDecryptor(Paths.get(keyPath));
        return decryptor.loadCredentials(credsPath);
    }
}

File: utilities/swift/Sources/CredentialCodeUtility/CredentialDecryptor.swift
import Foundation
import CryptoKit

/// A utility class for decrypting credentials from .creds files
public struct CredentialDecryptor {
    private let key: SymmetricKey
    
    /// Initialize with a base64-encoded key string
    public init(keyBase64: String) throws {
        guard let keyData = Data(base64Encoded: keyBase64.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            throw CredentialError.invalidKey
        }
        self.key = SymmetricKey(data: keyData)
    }
    
    /// Initialize with key data
    public init(keyData: Data) {
        self.key = SymmetricKey(data: keyData)
    }
    
    /// Initialize by reading key from file
    public init(keyPath: String) throws {
        let keyString = try String(contentsOfFile: keyPath, encoding: .utf8)
        try self.init(keyBase64: keyString)
    }
    
    /// Load and decrypt all credentials from a .creds file
    public func loadCredentials(from credsPath: String) throws -> [String: String] {
        let credsData = try Data(contentsOf: URL(fileURLWithPath: credsPath))
        return try loadCredentials(from: credsData)
    }
    
    /// Load and decrypt all credentials from .creds data
    public func loadCredentials(from credsData: Data) throws -> [String: String] {
        guard let credsJSON = try JSONSerialization.jsonObject(with: credsData) as? [String: Any],
              let credentials = credsJSON["credentials"] as? [String: [String: String]] else {
            throw CredentialError.invalidCredsFormat
        }
        
        var decryptedCredentials: [String: String] = [:]
        
        for (key, encryptedData) in credentials {
            decryptedCredentials[key] = try decrypt(encryptedData)
        }
        
        return decryptedCredentials
    }
    
    /// Decrypt a single credential
    public func decrypt(_ encryptedData: [String: String]) throws -> String {
        // Decode from base64
        guard let ciphertext = Data(base64Encoded: encryptedData["data"] ?? ""),
              let nonceData = Data(base64Encoded: encryptedData["nonce"] ?? ""),
              let tagData = Data(base64Encoded: encryptedData["tag"] ?? "") else {
            throw CredentialError.invalidData
        }
        
        // Create nonce and sealed box
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tagData)
        
        // Decrypt
        let decryptedData = try AES.GCM.open(sealedBox, using: key)
        
        guard let result = String(data: decryptedData, encoding: .utf8) else {
            throw CredentialError.decryptionFailed
        }
        
        return result
    }
}

/// Errors that can occur during credential operations
public enum CredentialError: LocalizedError {
    case invalidKey
    case invalidData
    case invalidCredsFormat
    case decryptionFailed
    
    public var errorDescription: String? {
        switch self {
        case .invalidKey:
            return "Invalid encryption key format"
        case .invalidData:
            return "Invalid encrypted data format"
        case .invalidCredsFormat:
            return "Invalid .creds file format"
        case .decryptionFailed:
            return "Failed to decrypt credential"
        }
    }
}